pkg/test-infra/tidb/operation.go

```diff
@@ -108,23 +111,36 @@ func (o *Ops) GetNodes() ([]cluster.Node, error) {
 func (o *Ops) GetClientNodes() ([]cluster.ClientNode, error) {
 	var clientNodes []cluster.ClientNode
 
-	ips, err := util.GetNodeIPs(o.cli, o.ns, map[string]string{"app.kubernetes.io/instance": o.name})
-	if err != nil {
-		return clientNodes, err
-	} else if len(ips) == 0 {
-		return clientNodes, errors.New("k8s node not found")
-	}
-
-	svc, err := o.getTiDBServiceByMeta(&o.tc.Service.ObjectMeta)
-	if err != nil {
-		return clientNodes, err
+	if util.IsInK8sPodEnvironment() {
+		svc, err := o.getTiDBServiceByClusterName(o.ns, o.name)
+		if err != nil {
+			return nil, err
+		}
+		clientNodes = append(clientNodes, cluster.ClientNode{
+			Namespace:   o.ns,
+			ClusterName: o.name,
+			IP:          svc.Spec.ClusterIP,
+			Port:        getTiDBServicePort(svc),
+		})
+	} else {
+		// If case isn't running on k8s pod, uses nodeIP:tidb_port as clientNode for conveniently debug on local
+		ips, err := util.GetNodeIPsFromPod(o.cli, o.ns, map[string]string{"app.kubernetes.io/instance": o.name})
+		if err != nil {
+			return nil, err
+		} else if len(ips) == 0 {
+			return nil, errors.New("k8s node not found")
+		}
+		svc, err := o.getTiDBServiceByClusterName(o.ns, o.name)
+		if err != nil {
+			return clientNodes, err
+		}
+		clientNodes = append(clientNodes, cluster.ClientNode{
```

Comment:
Maybe it's off-topic, but it seems that `GetClientNodes()` will only return one `ClientNode` in an array. Is this for backward compatibility?

Reply:
The concept `clientNode` is borrowed from Jepsen but there are a few differences against Jepsen.

In Jepsen, each client connects to different tidb-servers(deployed on docker-swarm), but because of the existence of service(as a load balancer) on K8s, the client can connect to tidb-servers through the service, so we simplify it to only one ClientNode on TiPocket.